arch/riscv/core/irq_manage.c

```diff
@@ -32,23 +34,28 @@ FUNC_NORETURN void z_irq_spurious(const void *unused)
 }
 
 #ifdef CONFIG_DYNAMIC_INTERRUPTS
-int arch_irq_connect_dynamic(unsigned int irq, unsigned int priority,
-			     void (*routine)(const void *parameter),
-			     const void *parameter, uint32_t flags)
+int arch_irq_intc_connect_dynamic(unsigned int irq, unsigned int priority,
+				  void (*routine)(const void *parameter),
+				  const void *parameter, uint32_t flags,
+				  const struct device *intc)
 {
-	ARG_UNUSED(flags);
-
-	z_isr_install(irq, routine, parameter);
-
 #if defined(CONFIG_RISCV_HAS_PLIC)
```

Comment:
Pardon my ignorance but won't this break the shared interrupt support on PLIC-based RISCV SoCs? `z_riscv_plic_isr_intc_install()` doesn't seem to be aware of the fact that the ISR may be set to `z_shared_isr()` and will overwrite it. Is this behavior desired?

Reply:
Yes I'm aware of this issue, the isr installer in the shared interrupt part has to be updated as well for this approach to work.

Reply:
> Pardon my ignorance but won't this break the shared interrupt support on PLIC-based RISCV SoCs? `z_riscv_plic_isr_intc_install()` doesn't seem to be aware of the fact that the ISR may be set to `z_shared_isr()` and will overwrite it. Is this behavior desired?

@LaurentiuM1234 - IIRC you recently updated the shared interrupt code. Were multi-level interrupts taken into account in the shared interrupt design?

If so, what would be the best way to support them?

Reply:
> @LaurentiuM1234 - IIRC you recently updated the shared interrupt code. Were multi-level interrupts taken into account in the shared interrupt design?

Yes the shared interrupt code is aware of the multi-level organization of the interrupts via `z_get_sw_isr_table_idx()` (for dynamic interrupts)


> If so, what would be the best way to support them?

Mm, not sure how helpful I can be here since I'm not familiar with RISCV and its quirks. If the multiple PLIC instances are basically aggregators for different interrupt lines I would suggest taking a look at https://github.com/zephyrproject-rtos/zephyr/pull/62776. The idea is that you register a dispatcher for each of these interrupt lines which will, in turn, call of the higher level reigstered ISRs. `z_isr_install()` is kept unchanged to keep the support for shared interrupts.